#!/usr/bin/env python3
"""
–¢–µ—Å—Ç –≥–µ–Ω–µ—Ä–∞—Ü–∏–∏ –ø—Ä–∏–≤–µ—Ç—Å—Ç–≤–µ–Ω–Ω–æ–≥–æ –∞—É–¥–∏–æ —á–µ—Ä–µ–∑ GenerateWelcomeAudio
"""
import sys
import asyncio
from pathlib import Path

# –î–æ–±–∞–≤–ª—è–µ–º –ø—É—Ç—å –∫ –º–æ–¥—É–ª—è–º
project_root = Path(__file__).parent.parent
sys.path.insert(0, str(project_root))

try:
    import grpc
    from grpc import aio
    
    # –ò–º–ø–æ—Ä—Ç protobuf —Ñ–∞–π–ª–æ–≤
    sys.path.insert(0, str(project_root / "modules" / "grpc_service"))
    import streaming_pb2
    import streaming_pb2_grpc
except ImportError as e:
    print(f"‚ùå –û—à–∏–±–∫–∞ –∏–º–ø–æ—Ä—Ç–∞: {e}")
    sys.exit(1)


async def test_generate_welcome_audio(host: str = "localhost", port: int = 50051):
    """–¢–µ—Å—Ç –≥–µ–Ω–µ—Ä–∞—Ü–∏–∏ –ø—Ä–∏–≤–µ—Ç—Å—Ç–≤–µ–Ω–Ω–æ–≥–æ –∞—É–¥–∏–æ"""
    print("=" * 60)
    print("–¢–µ—Å—Ç GenerateWelcomeAudio")
    print("=" * 60)
    print(f"–ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ {host}:{port}")
    print()
    
    # –°–æ–∑–¥–∞–µ–º –∫–∞–Ω–∞–ª
    address = f"{host}:{port}"
    channel = aio.insecure_channel(address)
    
    try:
        # –°–æ–∑–¥–∞–µ–º stub
        stub = streaming_pb2_grpc.StreamingServiceStub(channel)
        
        # –ñ–¥–µ–º –≥–æ—Ç–æ–≤–Ω–æ—Å—Ç–∏ –∫–∞–Ω–∞–ª–∞
        try:
            await asyncio.wait_for(channel.channel_ready(), timeout=5.0)
            print("‚úÖ –ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ —É—Å—Ç–∞–Ω–æ–≤–ª–µ–Ω–æ")
        except asyncio.TimeoutError:
            print("‚ùå –¢–∞–π–º–∞—É—Ç –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è")
            return False
        
        # –°–æ–∑–¥–∞–µ–º –∑–∞–ø—Ä–æ—Å
        request = streaming_pb2.WelcomeRequest(
            text="Hello! This is a test of audio generation. How are you today?",
            session_id="test_welcome_123"
        )
        
        print(f"üì§ –û—Ç–ø—Ä–∞–≤–∫–∞ –∑–∞–ø—Ä–æ—Å–∞:")
        print(f"   Text: {request.text}")
        print(f"   Session ID: {request.session_id}")
        print()
        
        # –í—ã–∑—ã–≤–∞–µ–º GenerateWelcomeAudio
        print("üéµ –ì–µ–Ω–µ—Ä–∞—Ü–∏—è –∞—É–¥–∏–æ...")
        chunk_count = 0
        total_bytes = 0
        error_occurred = False
        
        try:
            async for response in stub.GenerateWelcomeAudio(request, timeout=30.0):
                # –ü—Ä–æ–≤–µ—Ä—è–µ–º —Ç–∏–ø –∫–æ–Ω—Ç–µ–Ω—Ç–∞
                if response.HasField("audio_chunk"):
                    chunk_count += 1
                    audio_data = response.audio_chunk.audio_data
                    total_bytes += len(audio_data)
                    print(f"   ‚úÖ Chunk #{chunk_count}: {len(audio_data)} bytes "
                          f"(sample_rate={response.audio_chunk.sample_rate}, "
                          f"channels={response.audio_chunk.channels}, "
                          f"dtype={response.audio_chunk.dtype})")
                
                elif response.HasField("metadata"):
                    print(f"   üìã Metadata: {response.metadata}")
                
                elif response.HasField("end_message"):
                    print(f"   ‚úÖ {response.end_message}")
                
                elif response.HasField("error_message"):
                    print(f"   ‚ùå –û—à–∏–±–∫–∞: {response.error_message}")
                    error_occurred = True
                    break
        
        except grpc.RpcError as e:
            print(f"   ‚ùå gRPC –æ—à–∏–±–∫–∞: {e.code()} - {e.details()}")
            error_occurred = True
        
        print()
        print("=" * 60)
        if error_occurred:
            print("‚ùå –¢–µ—Å—Ç –Ω–µ –ø—Ä–æ–π–¥–µ–Ω: –ø—Ä–æ–∏–∑–æ—à–ª–∞ –æ—à–∏–±–∫–∞")
            return False
        elif chunk_count == 0:
            print("‚ö†Ô∏è –¢–µ—Å—Ç –Ω–µ –ø—Ä–æ–π–¥–µ–Ω: –Ω–µ –ø–æ–ª—É—á–µ–Ω–æ –∞—É–¥–∏–æ chunks")
            return False
        else:
            print(f"‚úÖ –¢–µ—Å—Ç –ø—Ä–æ–π–¥–µ–Ω —É—Å–ø–µ—à–Ω–æ!")
            print(f"   –ü–æ–ª—É—á–µ–Ω–æ chunks: {chunk_count}")
            print(f"   –í—Å–µ–≥–æ –±–∞–π—Ç: {total_bytes}")
            return True
        
    finally:
        await channel.close()


async def main():
    """–û—Å–Ω–æ–≤–Ω–∞—è —Ñ—É–Ω–∫—Ü–∏—è"""
    import argparse
    
    parser = argparse.ArgumentParser(description="–¢–µ—Å—Ç GenerateWelcomeAudio")
    parser.add_argument("--host", default="localhost", help="–•–æ—Å—Ç —Å–µ—Ä–≤–µ—Ä–∞")
    parser.add_argument("--port", type=int, default=50051, help="–ü–æ—Ä—Ç —Å–µ—Ä–≤–µ—Ä–∞")
    
    args = parser.parse_args()
    
    success = await test_generate_welcome_audio(args.host, args.port)
    sys.exit(0 if success else 1)


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n‚ö†Ô∏è –ü—Ä–µ—Ä–≤–∞–Ω–æ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª–µ–º")
        sys.exit(1)
    except Exception as e:
        print(f"‚ùå –ö—Ä–∏—Ç–∏—á–µ—Å–∫–∞—è –æ—à–∏–±–∫–∞: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)
